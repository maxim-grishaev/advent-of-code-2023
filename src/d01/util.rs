pub fn take_line_digits(line: &str) -> u32 {
    let mut l: u32 = 10;
    let mut r: u32 = 10;
    // iterate over the line's chars
    for (_i, c) in line.chars().enumerate() {
        let num = match c.to_digit(10) {
            None => continue,
            Some(n) => n,
        };

        if l == 10 {
            l = num;
            r = num;
        } else {
            r = num;
        }
    }
    return (l * 10 + r) as u32;
}
