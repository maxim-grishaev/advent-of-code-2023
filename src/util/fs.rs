use std::path::Path;

pub fn read_file(cur_file: &str, sibling_path: &str) -> String {
    let cur_dir = Path::new(cur_file).parent().unwrap();
    let ref input_path = cur_dir.join(sibling_path);
    // println!("file: {}", input_path.clone().to_str().unwrap());
    return std::fs::read_to_string(input_path)
        .expect("Something went wrong reading the file {cur_file}/{sibling_path}");
}
