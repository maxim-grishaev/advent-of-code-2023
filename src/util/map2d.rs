#![allow(unused)]

use std::fmt::{Debug, Formatter, Result};

#[derive(Debug)]
pub struct Coord2D {
    pub row: usize,
    pub col: usize,
}

impl Coord2D {
    pub fn new(row: usize, col: usize) -> Coord2D {
        Coord2D { row, col }
    }

    pub fn cell_distance(&self, other: &Coord2D) -> usize {
        (self.col as isize - other.col as isize).unsigned_abs()
            + (self.row as isize - other.row as isize).unsigned_abs()
    }
}

#[derive(Debug)]
pub struct Size {
    pub cols_count: usize,
    pub rows_count: usize,
}

#[derive(Debug)]
pub struct Lines2D {
    pub rows: Vec<usize>,
    pub cols: Vec<usize>,
}

pub type Vec2D<T> = Vec<Vec<T>>;

// #[derive(Debug)]
pub struct Map2D<T> {
    pub raw: Vec2D<T>,
}

const NEW_LINE: &str = "\n";
impl Debug for Map2D<char> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        let s = self
            .raw
            .iter()
            .map(|line| line.iter().collect::<String>())
            .collect::<Vec<String>>()
            .join(NEW_LINE);
        write!(f, "Map2D<char>\n{}", s)
    }
}

impl Map2D<char> {
    pub fn from_str(str: &str) -> Map2D<char> {
        Map2D::new(
            str.trim()
                .lines()
                .map(|line| line.chars().collect())
                .collect(),
        )
    }
}

impl<T> Map2D<T> {
    pub fn new(data: Vec<Vec<T>>) -> Map2D<T> {
        Map2D { raw: data }
    }

    pub fn get_val_at(&self, y_row: usize, x_col: usize) -> Option<&T> {
        self.raw.get(y_row).and_then(|line| line.get(x_col))
    }

    pub fn get_val_by_coord(&self, coord: &Coord2D) -> Option<&T> {
        self.get_val_at(coord.row, coord.col)
    }

    pub fn size(&self) -> Size {
        let height = self.raw.len();
        let width = self.raw.get(0).map(|line| line.len()).unwrap_or(0);
        Size {
            rows_count: height,
            cols_count: width,
        }
    }

    pub fn find_coords<F>(&self, is_found: F) -> Vec<Coord2D>
    where
        F: Fn(&T) -> bool,
    {
        let mut coords = vec![];
        self.for_each(|item, coord| {
            if is_found(item) {
                coords.push(coord);
            }
        });
        coords
    }

    pub fn for_each<F>(&self, mut clbk: F)
    where
        F: FnMut(&T, Coord2D),
    {
        for (y, row) in self.raw.iter().enumerate() {
            for (x, item) in row.iter().enumerate() {
                clbk(item, Coord2D { col: x, row: y });
            }
        }
    }

    pub fn find_lines_has_all_of<F>(&self, is_good: F) -> Lines2D
    where
        F: Fn(&T, Coord2D) -> bool,
    {
        let size = self.size();
        find_lines(size.rows_count, size.cols_count, |coord| {
            self.get_val_by_coord(&coord)
                .map_or(false, |c| is_good(c, coord))
        })
    }
}

fn bools_to_indexes(v: Vec<bool>) -> Vec<usize> {
    v.into_iter()
        .enumerate()
        .filter_map(|(idx, b)| match b {
            true => Some(idx),
            false => None,
        })
        .collect()
}

#[allow(clippy::needless_range_loop)]
fn find_lines<F>(rows_count: usize, cols_count: usize, is_found: F) -> Lines2D
where
    F: Fn(Coord2D) -> bool,
{
    let mut rows = vec![true; rows_count];
    let mut cols = vec![true; cols_count];
    for y in 0..rows_count {
        for x in 0..cols_count {
            if !is_found(Coord2D { row: y, col: x }) {
                rows[y] = false;
                cols[x] = false;
            }
        }
    }

    Lines2D {
        rows: bools_to_indexes(rows),
        cols: bools_to_indexes(cols),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn for_each() {
        let m2d = Map2D::new(vec![vec![1, 2], vec![3, 4]]);
        let mut sum = 0;
        m2d.for_each(|item, _coord| {
            sum += item;
        });
        assert_eq!(sum, 10);
    }

    const TEST_MAP_TINY: &str = r#"
..
1.
"#;
    #[test]
    fn find_lines_tiny() {
        let map2d = Map2D::from_str(TEST_MAP_TINY);
        println!("{:?}", map2d);
        let lines = map2d.find_lines_has_all_of(|c, _| c == &'.');
        assert_eq!(lines.rows, vec![0], "rows");
        assert_eq!(lines.cols, vec![1], "cols");
    }

    const TEST_MAP: &str = r#"
.....
1....
..2..
.....
....3
"#;
    #[test]
    fn find_lines() {
        let map2d = Map2D::from_str(TEST_MAP);
        let lines = map2d.find_lines_has_all_of(|c, _| c == &'.');
        assert_eq!(lines.rows, vec![0, 3], "rows");
        assert_eq!(lines.cols, vec![1, 3], "cols");
    }

    #[test]
    fn get_val_at() {
        let m2d = Map2D::new(vec![vec![1, 2], vec![3, 4]]);
        assert_eq!(m2d.get_val_at(0, 0), Some(&1));
        assert_eq!(m2d.get_val_at(0, 1), Some(&2));
        assert_eq!(m2d.get_val_at(1, 0), Some(&3));
        assert_eq!(m2d.get_val_at(1, 1), Some(&4));
        assert_eq!(m2d.get_val_at(2, 0), None);
        assert_eq!(m2d.get_val_at(0, 2), None);
    }
}
