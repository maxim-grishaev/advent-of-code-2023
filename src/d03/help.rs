use crate::util;

#[derive(Debug)]
pub struct NumPos {
    line_no: usize,
    col_no: usize,
    pub num: u32,
    digits_count: usize,
    // raw: Vec<char>,
}

fn val(line_no: usize, col_no: usize, num: Vec<char>) -> NumPos {
    NumPos {
        line_no: line_no.clone(),
        col_no: col_no.clone(),
        num: num.iter().collect::<String>().parse::<u32>().unwrap(),
        digits_count: num.len(),
        // raw: num,
    }
}

pub fn get_char_at(schema: &Vec<&str>, line_no: &isize, col_no: &isize) -> Option<char> {
    if line_no < &0 || col_no < &0 || *line_no >= schema.len() as isize {
        return None;
    }
    return schema[*line_no as usize].chars().nth(*col_no as usize);
}

pub fn is_adjacent(num: &NumPos, gear_coords: &(usize, usize)) -> bool {
    let distance = (num.line_no as isize - gear_coords.0 as isize).abs();
    if distance > 1 {
        return false;
    }
    let adjacent_coords = get_adjacent_coords(num);
    adjacent_coords.iter().any(|(line_no, col_no)| {
        *line_no == gear_coords.0 as isize && *col_no == gear_coords.1 as isize
    })
}

pub fn get_adjacent_coords(np: &NumPos) -> Vec<(isize, isize)> {
    let col_left: isize = np.col_no as isize - np.digits_count as isize - 1;
    let col_right = np.col_no as isize;

    let line_top = np.line_no as isize - 1;
    let line_cur = (np.line_no) as isize;
    let line_btm = (np.line_no + 1) as isize;

    let mut adjacent_coords: Vec<(isize, isize)> = vec![];

    // left vertical
    adjacent_coords.push((line_top, col_left));
    adjacent_coords.push((line_cur, col_left));
    adjacent_coords.push((line_btm, col_left));

    // horisontal
    (0..np.digits_count).for_each(|i| {
        adjacent_coords.push((line_top, col_left + 1 + i as isize));
        adjacent_coords.push((line_btm, col_left + 1 + i as isize));
    });

    // right vertical
    adjacent_coords.push((line_top, col_right));
    adjacent_coords.push((line_cur, col_right));
    adjacent_coords.push((line_btm, col_right));

    adjacent_coords
}

fn is_symbol(c: char) -> bool {
    !c.is_numeric() && c != '.'
}

pub fn is_detail(schema: &Vec<&str>, np: &NumPos) -> bool {
    get_adjacent_coords(np).iter().any(|(line_no, col_no)| {
        match get_char_at(schema, line_no, col_no) {
            Some(c) => is_symbol(c),
            None => return false,
        }
    })
}

pub fn find_all_numbers(schema: &Vec<&str>) -> Vec<NumPos> {
    let mut nums: Vec<NumPos> = vec![];
    loop {
        let (line_no, col_no) = match nums.last() {
            Some(last) => (last.line_no, last.col_no),
            None => (0, 0),
        };
        match find_number(schema, line_no, col_no) {
            Some(found) => {
                nums.push(found);
                continue;
            }
            None => return nums,
        }
    }
}

pub fn find_number(
    schema: &Vec<&str>,
    line_no_start: usize,
    col_no_start: usize,
) -> Option<NumPos> {
    if line_no_start >= schema.len() {
        return None;
    }
    let mut line_no = line_no_start;
    let mut col_no = col_no_start;
    let mut num: Vec<char> = vec![];

    // println!("start_pos: {:?} {:?}", line_no, col_no);

    let mut cur_line = schema[line_no].chars();
    // println!("line: {:?}", cur_line);
    if col_no > 0 {
        cur_line.nth(col_no - 1);
    }

    // println!("schema: {:?}", schema);
    // println!("cur_line: {:?}", cur_line);
    loop {
        // println!("loop: {:?} {:?} {:?}", line_no, col_no, cur_line);
        match cur_line.next() {
            Some(c) => {
                if c.is_numeric() {
                    num.push(c);
                    col_no += 1;
                    continue;
                }
                // end of number
                if num.len() > 0 {
                    // println!("end of number. num: {:?} @ {:?}", num, line_no);
                    return Some(val(line_no, col_no, num));
                }
                col_no += 1;
            }
            // end of line
            None => {
                if num.len() > 0 {
                    // println!("end of line. num: {:?} @ {:?}", num, line_no);
                    return Some(val(line_no, col_no, num));
                }

                // goto next line
                line_no += 1;
                col_no = 0;

                // end of schema?
                if line_no >= schema.len() {
                    return None;
                }
                // println!("NEXT LINE!");
                cur_line = schema[line_no].chars();
                // println!("line: {:?}", cur_line);
            }
        }
    }
    // None
}

pub fn read_schema() -> String {
    util::fs::read_file(file!(), "engine.txt")
}

#[cfg(test)]
mod test {
    use super::*;
    use insta::assert_debug_snapshot;

    const TEST_INPUT: &str = r#"
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
    "#;

    #[test]
    fn find_all() {
        let schema: Vec<&str> = TEST_INPUT.trim().lines().collect();
        let nums = find_all_numbers(&schema);
        assert_debug_snapshot!(nums);
    }

    #[test]
    fn get_adjacent_coords_snapshot() {
        let schema: Vec<&str> = TEST_INPUT.trim().lines().collect();
        let npo = find_number(&schema, 0, 0);
        let np = npo.unwrap();
        println!("{:?}", np);
        assert_debug_snapshot!(get_adjacent_coords(&np));
    }
}
