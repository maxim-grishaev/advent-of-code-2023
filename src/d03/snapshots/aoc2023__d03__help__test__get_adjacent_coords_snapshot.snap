---
source: src/d03/help.rs
expression: get_adjacent_coords(&np)
---
[
    (
        -1,
        -1,
    ),
    (
        0,
        -1,
    ),
    (
        1,
        -1,
    ),
    (
        -1,
        0,
    ),
    (
        1,
        0,
    ),
    (
        -1,
        1,
    ),
    (
        1,
        1,
    ),
    (
        -1,
        2,
    ),
    (
        1,
        2,
    ),
    (
        -1,
        3,
    ),
    (
        0,
        3,
    ),
    (
        1,
        3,
    ),
]
